processing/src/main/java/io/druid/query/GroupByMergedQueryRunner.java

```diff
@@ -79,7 +81,108 @@ public GroupByParallelQueryRunner(
   }
 
   @Override
-  public Sequence<T> run(final Query<T> queryParam, final Map<String, Object> responseContext)
+  public Sequence<T> run(final Query<T> query, final Map<String, Object> responseContext)
+  {
+    boolean isSingleThreaded = query.getContextValue(CTX_KEY_IS_SINGLE_THREADED, configSupplier.get().isSingleThreaded());
+    if (isSingleThreaded) {
+      return runSingleThreaded(query, responseContext);
+    } else {
+      return runParallel(query, responseContext);
+    }
+  }
+
+  private Sequence<T> runSingleThreaded(final Query<T> queryParam, final Map<String, Object> responseContext)
+  {
+    return new ConcatQueryRunner<>(
+        Sequences.map(
+            Sequences.simple(queryables),
+            new Function<QueryRunner<T>, QueryRunner<T>>()
+            {
+              @Override
+              public QueryRunner<T> apply(final QueryRunner<T> input)
+              {
+                return new QueryRunner<T>()
+                {
+                  @Override
+                  public Sequence<T> run(final Query<T> query, final Map<String, Object> responseContext)
+                  {
+                    final Pair<IncrementalIndex, Accumulator<IncrementalIndex, T>> indexAccumulatorPair = GroupByQueryHelper
```

Comment:
@himanshug there's redundancy in the getting some of the query flags and creating accumulators, not sure if that can be made better. I'm only harping on this because you were so adamant about similar redundancies in #2848 even though that code already existed before, just like in your case :)